network/src/services/dns_seeding/mod.rs

```diff
@@ -0,0 +1,161 @@
+            if network_state.with_peer_store(|peer_store| peer_store.random_peers(1).is_empty()) {
+                info!(target: "network", "No peer in peer store, start seeding...");
+                Instant::now()
+            } else {
+                Instant::now() + Duration::from_secs(11)
+            };
+        let check_interval = Interval::new_interval(Duration::from_secs(1));
+        DnsSeedingService {
+            network_state,
+            wait_until,
+            check_interval,
+            seeds,
+        }
+    }
+
+    fn seeding(&self) -> Result<(), Box<dyn Error>> {
+        let enough_outbound = self.network_state.with_peer_registry(|reg| {
+            reg.peers()
+                .values()
+                .filter(|peer| peer.is_outbound())
+                .count()
+                >= 2
+        });
+        if enough_outbound {
+            debug!(target: "network", "Enough outbound peers");
+            return Ok(());
+        }
+
+        let mut pubkey_bytes = [4u8; 65];
+        hex_decode(TXT_VERIFY_PUBKEY.as_bytes(), &mut pubkey_bytes[1..65])
+            .map_err(|err| format!("parse key({}) error: {:?}", TXT_VERIFY_PUBKEY, err))?;
+        let pubkey = PublicKey::from_slice(&pubkey_bytes)
+            .map_err(|err| format!("create PublicKey failed: {:?}", err))?;
+
+        let resolver = DnsConfig::load_default()
+            .map_err(|err| format!("Failed to load system configuration: {}", err))
+            .and_then(|config| {
+                DnsResolver::new(config)
+                    .map_err(|err| format!("Failed to create DNS resolver: {}", err))
+            })?;
+
+        let mut addrs = Vec::new();
+        for seed in &self.seeds {
+            debug!(target: "network", "query txt records from: {}", seed);
+            match resolver.resolve_record::<Txt>(seed) {
+                Ok(records) => {
+                    for record in records {
+                        match std::str::from_utf8(&record.data) {
+                            Ok(record) => match SeedRecord::decode_with_pubkey(&record, &pubkey) {
+                                Ok(seed_record) => {
+                                    let address = seed_record.address();
+                                    trace!(target: "network", "got dns txt address: {}", address);
+                                    addrs.push(address);
+                                }
+                                Err(err) => {
+                                    debug!(target: "network", "decode dns txt record failed: {:?}, {:?}", err, record);
+                                }
+                            },
+                            Err(err) => {
+                                debug!(target: "network", "get dns txt record error: {:?}", err);
+                            }
+                        }
+                    }
+                }
+                Err(_) => {
+                    if let Ok(addr) = seed.to_multiaddr() {
```

Comment:
DNS may fail temporarily. Can we explicitly differentiate between a peer address and a DNS in `bootnodes`, such as using a special format so that DNS domain should not be recognized as a valid multiaddr.

Reply:
OK